File: vk/python/verify_local.py
# Локальная проверка VK echo-бота БЕЗ реального токена и БЕЗ обращения к проду VK.
# Поднимаем фейковый VK-сервер (aiohttp), направляем НАСТОЯЩИЙ SDK на него (api.API_URL),
# симулируем Bots Long Poll и ловим исходящие messages.send (это и есть ответы бота).
# Хендлеры берём из bot.py напрямую (from bot import labeler) — гоняем реальные обработчики.
import asyncio
import shutil
import socket
import sys

from aiohttp import web

from vkbottle import API
from vkbottle.bot import Bot

from bot import GREETING, labeler  # реальные хендлеры on_start + echo, зарегистрированные на labeler

GROUP_ID = 42
ECHO_PEER = 2000000001
START_PEER = 2000000002
PROBE_PEER = 2000000003
ECHO_TEXT = "Привет, VK!"
TS = 1700000000

# client_info обязателен в каждом апдейте (иначе MessageMin не собирается).
CLIENT_INFO = {
    "button_actions": ["text"],
    "keyboard": True,
    "inline_keyboard": True,
    "carousel": False,
    "lang_id": 0,
}

posts = []            # пойманные messages.send: {peer_ids, text, access_token, v}
batch_delivered = False


def _message(msg_id, peer_id, from_id, text, payload=None):
    # Обязательные для vkbottle MessageMin поля + ловушка fwd_messages: [].
    msg = {
        "id": msg_id,
        "date": TS,
        "peer_id": peer_id,
        "from_id": from_id,
        "text": text,
        "conversation_message_id": msg_id,
        "out": 0,
        "version": 0,
        "fwd_messages": [],
        "attachments": [],
    }
    if payload is not None:
        msg["payload"] = payload
    return msg


def _update(msg_id, peer_id, from_id, text, payload=None):
    return {
        "type": "message_new",
        "object": {
            "client_info": CLIENT_INFO,
            "message": _message(msg_id, peer_id, from_id, text, payload),
        },
        "group_id": GROUP_ID,
    }


def make_app(base_url):
    async def handle_method(request):
        method = request.match_info["method"]
        if method == "groups.getById":
            return web.json_response({"response": {"groups": [{"id": GROUP_ID}]}})
        if method == "groups.getLongPollServer":
            # server используется дословно как URL → замыкаем long poll на себя
            return web.json_response({"response": {"key": "k", "server": f"{base_url}/longpoll", "ts": "1"}})
        if method == "messages.send":
            form = await request.post()                # параметры метода — form-data
            posts.append({
                "peer_ids": form.get("peer_ids"),
                "text": form.get("message"),
                "access_token": request.query.get("access_token"),  # авторизация — в query
                "v": request.query.get("v"),
            })
            print(f"[mock] ← messages.send peer_ids={form.get('peer_ids')} text={form.get('message')!r}")
            peer = int(form.get("peer_ids")) if form.get("peer_ids") else 0
            # при peer_ids SDK ждёт СПИСОК
            return web.json_response({"response": [{"peer_id": peer, "message_id": 555, "conversation_message_id": 7}]})
        return web.json_response({"response": 1})      # прочие методы — безобидный дефолт

    async def handle_longpoll(request):
        global batch_delivered
        if not batch_delivered:
            batch_delivered = True
            print("[mock] → отдаю 3 апдейта: текст, «Начать»+payload, probe(text='')")
            return web.json_response({"ts": "2", "updates": [
                _update(1, ECHO_PEER, 101, ECHO_TEXT),
                _update(2, START_PEER, 102, "Начать", payload='{"command": "start"}'),
                _update(3, PROBE_PEER, 103, ""),  # attachment-only: text="" → ответа быть не должно
            ]})
        await asyncio.sleep(0.3)  # эмуляция long-poll: пусто с задержкой (anti-busy-loop + быстрый stop)
        return web.json_response({"ts": "2", "updates": []})

    app = web.Application()
    app.router.add_post("/method/{method}", handle_method)
    app.router.add_post("/longpoll", handle_longpoll)
    return app


async def main() -> int:
    # пред-биндим сокет, чтобы знать порт до старта (как в max/python)
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind(("127.0.0.1", 0))
    port = sock.getsockname()[1]
    base_url = f"http://127.0.0.1:{port}"

    runner = web.AppRunner(make_app(base_url))
    await runner.setup()
    site = web.SockSite(runner, sock)
    await site.start()
    print(f"[mock] фейковый VK слушает {base_url}\n")

    # НАСТОЯЩИЙ SDK, направленный на мок. API_URL — инстанс-атрибут перекрывает классовый;
    # ОБЯЗАТЕЛЕН хвост /method/ со слэшем (код делает конкатенацию API_URL + method).
    api = API("mock-token")
    api.API_URL = f"{base_url}/method/"
    bot = Bot(api=api, labeler=labeler)

    polling_task = asyncio.create_task(bot.run_polling())

    # ждём 2 ожидаемых messages.send (эхо + приветствие) максимум ~10 с
    for _ in range(100):
        if len(posts) >= 2:
            break
        await asyncio.sleep(0.1)
    await asyncio.sleep(1.0)  # шанс на возможный (нежелательный) ответ на probe

    bot.polling.stop()
    polling_task.cancel()
    try:
        await polling_task
    except asyncio.CancelledError:
        pass
    await runner.cleanup()
    shutil.rmtree(".vkbottle", ignore_errors=True)  # BotPolling пишет ./.vkbottle/bot-polling/{id}.json

    # --- проверки (зеркало max/python) ---
    echo_post = next((p for p in posts if p["peer_ids"] == str(ECHO_PEER)), None)
    start_post = next((p for p in posts if p["peer_ids"] == str(START_PEER)), None)

    text_ok = echo_post is not None and echo_post["text"] == ECHO_TEXT
    start_ok = start_post is not None and start_post["text"] == GREETING
    auth_ok = len(posts) > 0 and all(p["access_token"] == "mock-token" and p["v"] == "5.199" for p in posts)
    probe_ok = len(posts) == 2 and not any(p["peer_ids"] == str(PROBE_PEER) for p in posts)

    print("\n=== РЕЗУЛЬТАТ ===")
    print(f"✅ эхо текста:        {(echo_post['text'] if echo_post else None)!r}  (ожидали {ECHO_TEXT!r})  {'✅' if text_ok else '❌'}")
    print(f"✅ «Начать» payload:  {(start_post['text'] if start_post else None)!r}  (приветствие, НЕ эхо «Начать»)  {'✅' if start_ok else '❌'}")
    print(f"✅ авторизация:       access_token=mock-token и v=5.199 в query  {'✅' if auth_ok else '❌'}")
    print(f"🔍 probe:             messages.send всего {len(posts)}  (ожидали 2; на text='' ответа нет)  {'✅' if probe_ok else '❌'}")

    ok = text_ok and start_ok and auth_ok and probe_ok
    print("\n✅ PASS: echo round-trip через настоящий SDK; payload-приветствие отдельным хендлером; пустой текст проигнорирован."
          if ok else "\n❌ FAIL: см. отметки выше.")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: vk/typescript/verify-local.ts
// Локальная проверка VK echo-бота БЕЗ реального токена и БЕЗ обращения к проду VK.
// Поднимаем фейковый VK-сервер (node:http), направляем НАСТОЯЩИЙ vk-io на него (apiBaseUrl),
// симулируем Bots Long Poll и ловим исходящие messages.send (это и есть ответы бота).
// Хендлеры берём из src/handlers.ts (registerHandlers) — гоняем реальную логику бота.
import { createServer, type IncomingMessage } from 'node:http';
import type { AddressInfo } from 'node:net';
import { VK } from 'vk-io';
import { registerHandlers, GREETING } from './src/handlers.js';

const GROUP_ID = 42;
const ECHO_PEER = 2000000001;
const START_PEER = 2000000002;
const PROBE_PEER = 2000000003;
const ECHO_TEXT = 'Привет, VK!';
const TS = 1700000000;

// client_info обязателен — без него object трактуется как голое сообщение и структура ломается.
const CLIENT_INFO = {
  button_actions: ['text'],
  keyboard: true,
  inline_keyboard: true,
  carousel: false,
  lang_id: 0,
};

interface Sent {
  peerIds: string | null;
  text: string | null;
  accessToken: string | null;
  v: string | null;
}
const sent: Sent[] = [];
let batchDelivered = false;
let port = 0;

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

function message(id: number, peerId: number, fromId: number, text: string | undefined, payload?: string) {
  // Для эхо-логики vk-io нужны: id, date, peer_id, from_id, out, conversation_message_id, attachments.
  // text/payload опциональны — их отсутствие = undefined в контексте (так моделируем probe без текста).
  const msg: Record<string, unknown> = {
    id,
    date: TS,
    peer_id: peerId,
    from_id: fromId,
    out: 0,
    conversation_message_id: id,
    attachments: [],
  };
  if (text !== undefined) msg.text = text;
  if (payload !== undefined) msg.payload = payload;
  return msg;
}

function update(id: number, peerId: number, fromId: number, text: string | undefined, payload?: string) {
  return {
    type: 'message_new',
    object: { client_info: CLIENT_INFO, message: message(id, peerId, fromId, text, payload) },
    group_id: GROUP_ID,
  };
}

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve) => {
    let data = '';
    req.on('data', (c) => (data += c));
    req.on('end', () => resolve(data));
  });
}

const server = createServer(async (req, res) => {
  const url = new URL(req.url ?? '', 'http://127.0.0.1');
  const json = (obj: unknown) => {
    res.writeHead(200, { 'content-type': 'application/json' });
    res.end(JSON.stringify(obj));
  };

  // groups.getById — vk-io резолвит group_id (читает groups[0].id)
  if (url.pathname === '/method/groups.getById') {
    return json({ response: { groups: [{ id: GROUP_ID }] } });
  }
  // getLongPollServer — server используется ДОСЛОВНО как URL → замыкаем long poll на себя
  if (url.pathname === '/method/groups.getLongPollServer') {
    return json({ response: { key: 'k', server: `http://127.0.0.1:${port}/longpoll`, ts: '1' } });
  }
  // long-poll — у vk-io это GET с query (act=a_check, ...); мок игнорирует параметры
  if (url.pathname === '/longpoll') {
    if (!batchDelivered) {
      batchDelivered = true;
      console.log('[mock] → отдаю 3 апдейта: текст, «Начать»+payload, probe(без text)');
      return json({
        ts: '2',
        updates: [
          update(1, ECHO_PEER, 101, ECHO_TEXT),
          update(2, START_PEER, 102, 'Начать', '{"command":"start"}'),
          update(3, PROBE_PEER, 103, undefined), // нет text → context.text === undefined → молчим
        ],
      });
    }
    await sleep(300); // эмуляция long-poll: пусто с задержкой (быстрый stop без зависания)
    return json({ ts: '2', updates: [] });
  }
  // messages.send — у vk-io access_token и v в form-ТЕЛЕ (не в query, в отличие от Python-соседа)
  if (url.pathname === '/method/messages.send') {
    const params = new URLSearchParams(await readBody(req));
    sent.push({
      peerIds: params.get('peer_ids'),
      text: params.get('message'),
      accessToken: params.get('access_token'),
      v: params.get('v'),
    });
    console.log(`[mock] ← messages.send peer_ids=${params.get('peer_ids')} text=${JSON.stringify(params.get('message'))}`);
    return json({
      response: [{ peer_id: Number(params.get('peer_ids')), message_id: 555, conversation_message_id: 7 }],
    });
  }
  json({ response: 1 }); // прочие методы — безобидный дефолт
});

await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
port = (server.address() as AddressInfo).port;
console.log(`[mock] фейковый VK слушает http://127.0.0.1:${port}\n`);

// НАСТОЯЩИЙ vk-io, направленный на мок. agent: undefined — иначе https.globalAgent
// не умеет http:// (ловушка). pollingGroupId НЕ задаём — пусть резолвится через groups.getById.
const vk = new VK({
  token: 'test-token',
  apiBaseUrl: `http://127.0.0.1:${port}/method`,
  agent: undefined,
});
registerHandlers(vk);
await vk.updates.start();

for (let i = 0; i < 100 && sent.length < 2; i++) await sleep(100); // ждём 2 send ~10 с
await sleep(800); // шанс на возможный (нежелательный) ответ на probe

await vk.updates.stop();
server.closeAllConnections();
await new Promise<void>((resolve) => server.close(() => resolve()));

// --- проверки (зеркало vk/python) ---
const echo = sent.find((s) => s.peerIds === String(ECHO_PEER));
const start = sent.find((s) => s.peerIds === String(START_PEER));

const textOk = echo?.text === ECHO_TEXT;
const startOk = start?.text === GREETING;
const authOk = sent.length > 0 && sent.every((s) => s.accessToken === 'test-token' && s.v === '5.199');
const probeOk = sent.length === 2 && !sent.some((s) => s.peerIds === String(PROBE_PEER));

console.log('\n=== РЕЗУЛЬТАТ ===');
console.log(`✅ эхо текста:        ${JSON.stringify(echo?.text)}  (ожидали ${JSON.stringify(ECHO_TEXT)})  ${textOk ? '✅' : '❌'}`);
console.log(`✅ «Начать» payload:  ${JSON.stringify(start?.text)}  (приветствие, НЕ эхо «Начать»)  ${startOk ? '✅' : '❌'}`);
console.log(`✅ авторизация:       access_token + v=5.199 в form-теле  ${authOk ? '✅' : '❌'}`);
console.log(`🔍 probe:             messages.send всего ${sent.length}  (ожидали 2; без text ответа нет)  ${probeOk ? '✅' : '❌'}`);

const ok = textOk && startOk && authOk && probeOk;
console.log(ok
  ? '\n✅ PASS: echo round-trip через настоящий SDK; payload-приветствие отдельной веткой; сообщение без текста проигнорировано.'
  : '\n❌ FAIL: см. отметки выше.');
process.exit(ok ? 0 : 1);
